# Machine Translation



# 1)- Importing key modules

In [1]:
#support both Python 2 and Python 3 with minimal overhead.
from __future__ import absolute_import, division, print_function
#ignore warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
import string 
from string import digits
from collections import Counter
import re 
from sklearn.utils import shuffle
import pandas as pd 
from pickle import dump
from pickle import load
from unicodedata import normalize
import matplotlib.pyplot as plt 
% matplotlib inline 
pd.set_option('display.max_colwidth', 200)

UsageError: Line magic function `%` not found.


# 2)- Loading data

We have data from 2009 to 2016.

In [3]:
# load doc into memory
def load_doc(filename):
	# open the file as read only
	file = open(filename, mode='rt', encoding='utf-8')
	# read all text
	text = file.read()
	# close the file
	file.close()
	return text

In [4]:
# split a loaded document into sentences
def to_sentences(doc):
	return doc.strip().split('\n')

In [5]:
# shortest and longest sentence lengths
def sentence_lengths(sentences):
	lengths = [len(s.split()) for s in sentences]
	return min(lengths), max(lengths)

### 2.1)- For year 2009

In [6]:
# load English data
filename = 'newstest2009.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2009.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=2525, min=1, max=108
German data: sentences=2525, min=1, max=110


It is important to notice that sentence length is same. So, we have a balanced data.

We shall check this on all years

### 2.2)-For 2010

In [7]:
# load English data
filename = 'newstest2010.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2010.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=2489, min=1, max=74
German data: sentences=2489, min=1, max=86


### 2.3)-For year 2011

In [8]:
# load English data
filename = 'newstest2011.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2011.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=3003, min=1, max=93
German data: sentences=3003, min=1, max=92


### 2.4)-For year 2012

In [9]:
# load English data
filename = 'newstest2012.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2012.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=3003, min=1, max=114
German data: sentences=3003, min=1, max=101


### 2.5)-For year 2013

In [10]:
# load English data
filename = 'newstest2013.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2013.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=3000, min=1, max=82
German data: sentences=3000, min=1, max=85


### 2.6)-For year 2014

In [11]:
# load English data
filename = 'newstest2014.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2014.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=3003, min=1, max=68
German data: sentences=3003, min=1, max=64


### 2.7)- For year 2015

In [12]:
# load English data
filename = 'newstest2015.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2015.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=2169, min=1, max=71
German data: sentences=2169, min=1, max=72


### For year 2016

In [13]:
# load English data
filename = 'newstest2016.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2016.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=2999, min=1, max=83
German data: sentences=2999, min=1, max=88


**All datasets have balanced sentences for English and German version**

# 2)- Combine Files

- We can do this whole process on each file and then use one dataset pair for modeling i.e German 2015, English 2015.

- Or we can combine all these files and then apply these above discussed preprocessing methods on combined data. 

-  We may end up with a huge data so, we need to use random sample of data to train our model.

- Finally, I personally like to work with dataframes and like to see files as form of matrix or vector. So, I will convert these files into dataframes then combine both English and German translation parts in rows and columns. So, they look agreeable to me. 

In [14]:
df_en_2009= pd.read_table('newstest2009.en', sep="\t", names=['eng'])
df_en_2009.shape

(2441, 1)

In [15]:
df_ger_2009= pd.read_table('newstest2009.de', sep="\t", names=['ger'])
df_ger_2009.shape

(2394, 1)

In [16]:
# load English data
filename = 'newstest2009.en'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('English data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

# load French data
filename = 'newstest2009.de'
doc = load_doc(filename)
sentences = to_sentences(doc)
minlen, maxlen = sentence_lengths(sentences)
print('German data: sentences=%d, min=%d, max=%d' % (len(sentences), minlen, maxlen))

English data: sentences=2525, min=1, max=108
German data: sentences=2525, min=1, max=110


**As pandas dataframes show different lengths. How can we combine these two files as one column shows english and other column shows german**

In [39]:
df_en_2009.head()

eng
0                                                                                  Prague Stock Market falls to minus by the end of the trading day
1                                                                  After a sharp drop in the morning, the Prague Stock Market corrected its losses.
2                                 Transactions with stocks from the Czech Energy Enterprise (ČEZ) reached nearly half of the regular daily trading.
3  The Prague Stock Market immediately continued its fall from Monday at the beginning of Tuesday's trading, when it dropped by nearly six percent.
4                                                                          This time the fall in stocks on Wall Street is responsible for the drop.

In [40]:
df_ger_2009.head()

ger
0                                                                                     Die Prager Börse stürzt gegen Geschäftsschluss ins Minus.
1                                                           Nach dem steilen Abfall am Morgen konnte die Prager Börse die Verluste korrigieren.
2                                              Die Transaktionen mit den Aktien von ČEZ erreichten fast die Hälfte des normalen Tagesgeschäfts.
3  Die Prager Börse knüpfte gleich zu Beginn der Dienstagsgeschäfte an den Einbruch vom Montag an, als sie um weitere sechs Prozentpunkte sank.
4                                                                                    Diesmal lag der Grund für den Einbruch an der Wall Street.

In [18]:
df_combine_2009=pd.concat([df_en_2009, df_ger_2009], axis=1, join='inner') # to adjust automatically row labels i.e index
df_combine_2009.head()

eng  \
0  Prague Stock Market falls to minus by the end ...   
1  After a sharp drop in the morning, the Prague ...   
2  Transactions with stocks from the Czech Energy...   
3  The Prague Stock Market immediately continued ...   
4  This time the fall in stocks on Wall Street is...   

                                                 ger  
0  Die Prager Börse stürzt gegen Geschäftsschluss...  
1  Nach dem steilen Abfall am Morgen konnte die P...  
2  Die Transaktionen mit den Aktien von ČEZ errei...  
3  Die Prager Börse knüpfte gleich zu Beginn der ...  
4  Diesmal lag der Grund für den Einbruch an der ...

In [19]:
df_combine_2009.shape

(2394, 2)

**We have lost data as english version had 2441 and german had 2394.Now we have 2394.**

# Problem:
   
**How could we combine these two data files without losing anything?**

**If we want to analyse random data after combining all files , will df.sample(5000) work ?**
    

### END

**Here we have combined one file. Let's us combine all of dataset in form of dataframes with columns (showing german and english lang text) and rows.**

### 5a)- Combine all english version

In [0]:
df_en_2010= pd.read_table('newstest2010.en', sep="\t", names=['eng'])
df_en_2011= pd.read_table('newstest2011.en', sep="\t", names=['eng'])
df_en_2012= pd.read_table('newstest2012.en', sep="\t", names=['eng'])
df_en_2013= pd.read_table('newstest2013.en', sep="\t", names=['eng'])
df_en_2014= pd.read_table('newstest2014.en', sep="\t", names=['eng'])
df_en_2015= pd.read_table('newstest2015.en', sep="\t", names=['eng'])
df_en_2016= pd.read_table('newstest2016.en', sep="\t", names=['eng'])

In [0]:
df_eng_all=pd.concat([df_en_2009,df_en_2010,df_en_2011,df_en_2012,df_en_2013,df_en_2014,df_en_2015,df_en_2016], ignore_index=True)

In [56]:
df_eng_all.head()

eng
0                                                                                  Prague Stock Market falls to minus by the end of the trading day
1                                                                  After a sharp drop in the morning, the Prague Stock Market corrected its losses.
2                                 Transactions with stocks from the Czech Energy Enterprise (ČEZ) reached nearly half of the regular daily trading.
3  The Prague Stock Market immediately continued its fall from Monday at the beginning of Tuesday's trading, when it dropped by nearly six percent.
4                                                                          This time the fall in stocks on Wall Street is responsible for the drop.

### 5b)- Combine all German version

In [0]:
df_ger_2010= pd.read_table('newstest2010.de', sep="\t", names=['ger'])
df_ger_2011= pd.read_table('newstest2011.de', sep="\t", names=['ger'])
df_ger_2012= pd.read_table('newstest2012.de', sep="\t", names=['ger'])
df_ger_2013= pd.read_table('newstest2013.de', sep="\t", names=['ger'])
df_ger_2014= pd.read_table('newstest2014.de', sep="\t", names=['ger'])
df_ger_2015= pd.read_table('newstest2015.de', sep="\t", names=['ger'])
df_ger_2016= pd.read_table('newstest2016.de', sep="\t", names=['ger'])

In [0]:
df_ger_all=pd.concat([df_ger_2009,df_ger_2010,df_ger_2011,df_ger_2012,df_ger_2013,df_ger_2014,df_ger_2015,df_ger_2016], ignore_index=True)

In [61]:
df_ger_all.head()

ger
0                                                                                     Die Prager Börse stürzt gegen Geschäftsschluss ins Minus.
1                                                           Nach dem steilen Abfall am Morgen konnte die Prager Börse die Verluste korrigieren.
2                                              Die Transaktionen mit den Aktien von ČEZ erreichten fast die Hälfte des normalen Tagesgeschäfts.
3  Die Prager Börse knüpfte gleich zu Beginn der Dienstagsgeschäfte an den Einbruch vom Montag an, als sie um weitere sechs Prozentpunkte sank.
4                                                                                    Diesmal lag der Grund für den Einbruch an der Wall Street.

### 5c)- Join english and german versions

In [64]:
# combine into rows and columns dataframe

df_combine=pd.concat([df_eng_all,df_ger_all], axis=1, join='inner') # to adjust automatically row labels i.e index
df_combine.head(15)

eng                                                                                                                                                                                                      ger
0                                                                                                                                          Prague Stock Market falls to minus by the end of the trading day                                                                                                                                                Die Prager Börse stürzt gegen Geschäftsschluss ins Minus.
1                                                                                                                          After a sharp drop in the morning, the Prague Stock Market corrected its losses.                                                                                                                      Nach dem steilen Abfall am Morgen konnte die Prager Börse die Verluste korrigieren.
2                                                                                         Transactions with stocks from the Czech Energy Enterprise (ČEZ) reached nearly half of the regular daily trading.                                                                                                         Die Transaktionen mit den Aktien von ČEZ erreichten fast die Hälfte des normalen Tagesgeschäfts.
3                                                          The Prague Stock Market immediately continued its fall from Monday at the beginning of Tuesday's trading, when it dropped by nearly six percent.                                                             Die Prager Börse knüpfte gleich zu Beginn der Dienstagsgeschäfte an den Einbruch vom Montag an, als sie um weitere sechs Prozentpunkte sank.
4                                                                                                                                  This time the fall in stocks on Wall Street is responsible for the drop.                                                                                                                                               Diesmal lag der Grund für den Einbruch an der Wall Street.
5   The reaction of the market to the results of the vote in the American House of Representatives, which refused to support the plan for the stabilization of the financial sector there, has manifeste...  Auch in Tschechien zeigt sich so die Reaktion des Marktes auf das Ergebnis der Abstimmung des amerikanischen Repräsentantenhauses, das es abgelehnt hatte, einen Stabilisierungsplan für den dortige...
6                                                                                                                                                                                       Stocks fall in Asia                                                                                                                                                                                 Fall der Aktien in Asien
7                                                        Stocks in the Asian markets experienced a dramatic drop on Tuesday, even though the indexes ultimately erased a part of the losses during the day.                                   Einen dramatischen Fall erlebten am Dienstag die Aktien an den Börsen in Asien, wenngleich die Indices im Laufe des Tages einen Teil der Verluste ausgleichen konnten.
8               The Hang Seng Index of the Hong Kong Stock Exchange wrote off nearly four percent during the day, but later it erased a part of the losses and reduced the decrease to roughly 2.5 percent.  Der Index der Hang Seng Börse in Hongkong verlor im Laufe des Geschäftstages fast vier Prozentpunkte, konnte aber später einen Teil der Verluste wettmachen, somit verringerte sich der Rückgang auf...
9                        The Hang Seng China Enterprises Index, which follows the movement of Chinese stocks on the stock market in Hong Kong, dropped

### 5d)- Saving file

In [0]:
df_combine.to_pickle('full_data.pkl')

In [0]:
# Load again to check
df=pd.read_pickle('full_data.pkl')

In [69]:
df.head(15)

eng                                                                                                                                                                                                      ger
0                                                                                                                                          Prague Stock Market falls to minus by the end of the trading day                                                                                                                                                Die Prager Börse stürzt gegen Geschäftsschluss ins Minus.
1                                                                                                                          After a sharp drop in the morning, the Prague Stock Market corrected its losses.                                                                                                                      Nach dem steilen Abfall am Morgen konnte die Prager Börse die Verluste korrigieren.
2                                                                                         Transactions with stocks from the Czech Energy Enterprise (ČEZ) reached nearly half of the regular daily trading.                                                                                                         Die Transaktionen mit den Aktien von ČEZ erreichten fast die Hälfte des normalen Tagesgeschäfts.
3                                                          The Prague Stock Market immediately continued its fall from Monday at the beginning of Tuesday's trading, when it dropped by nearly six percent.                                                             Die Prager Börse knüpfte gleich zu Beginn der Dienstagsgeschäfte an den Einbruch vom Montag an, als sie um weitere sechs Prozentpunkte sank.
4                                                                                                                                  This time the fall in stocks on Wall Street is responsible for the drop.                                                                                                                                               Diesmal lag der Grund für den Einbruch an der Wall Street.
5   The reaction of the market to the results of the vote in the American House of Representatives, which refused to support the plan for the stabilization of the financial sector there, has manifeste...  Auch in Tschechien zeigt sich so die Reaktion des Marktes auf das Ergebnis der Abstimmung des amerikanischen Repräsentantenhauses, das es abgelehnt hatte, einen Stabilisierungsplan für den dortige...
6                                                                                                                                                                                       Stocks fall in Asia                                                                                                                                                                                 Fall der Aktien in Asien
7                                                        Stocks in the Asian markets experienced a dramatic drop on Tuesday, even though the indexes ultimately erased a part of the losses during the day.                                   Einen dramatischen Fall erlebten am Dienstag die Aktien an den Börsen in Asien, wenngleich die Indices im Laufe des Tages einen Teil der Verluste ausgleichen konnten.
8               The Hang Seng Index of the Hong Kong Stock Exchange wrote off nearly four percent during the day, but later it erased a part of the losses and reduced the decrease to roughly 2.5 percent.  Der Index der Hang Seng Börse in Hongkong verlor im Laufe des Geschäftstages fast vier Prozentpunkte, konnte aber später einen Teil der Verluste wettmachen, somit verringerte sich der Rückgang auf...
9                        The Hang Seng China Enterprises Index, which follows the movement of Chinese stocks on the stock market in Hong Kong, dropped

Now we are ready for next stage i.e modeling